Upsell: base template.

nicolad opened the pull request:
Related: https://github.com/eventespresso/barista/issues/723

nicolad commented on packages/ee-components/src/Upsell/templates/BaseTemplate.tsx:
```diff
@@ -0,0 +1,45 @@
+import classNames from 'classnames';
+
+import { Image } from '@eventespresso/adapters';
+import { Button, Heading } from '@eventespresso/ui-components';
```
Initially I was using raw `h1`/ `h2` and was thinking to go with this route in order to make upsell component different than our consolidated style, but later realized that it's better to use from existing component and, maybe, extend those, because we have important `a11y` improvements we might need to duplicate (not in case of heading, but in case of other components)

tn3rb replied:
agreed 💯

tn3rb commented on packages/ee-components/src/Upsell/Upsell.tsx:
```diff
@@ -1,29 +1,16 @@
 
-		return null;
-	};
+	if (template === 'base') {
+		return <BaseTemplate {...props} className={className} />;
+	}
```
another option would be to perform some string manipulation and use the template prop as the component name, ie:

```tsx
const Component = capitalize(template) + 'Template';
return <Component {...props} className={className} />;
```

that is... **IF** that is possible (i think we've done something like this already somewhere).
Would require loading ALL of the templates at the top of the file (which we would have to do regardless)

nicolad replied:
we have that approach with regards to mutations

![image](https://user-images.githubusercontent.com/1477453/108902826-41847780-7625-11eb-91d9-133907c8ec2f.png)

nicolad replied:
remind me please where exactly that approach is already being used

nicolad replied:
@manzoorwanijk gave some more info to this implementation:
> We had this in core before barista
No, we simply didn't port that over
Also, it was a bit too complex with workspaces

***
![image](https://user-images.githubusercontent.com/1477453/109005778-02046c80-76b3-11eb-963c-966010190972.png)

manzoorwanijk replied:
Well, in this case, dynamic imports may be a better idea, or as @tn3rb mentioned, creating an object of all the templates and then loading the correct template based upon the template name/id.

tn3rb commented on packages/ee-components/src/Upsell/style.scss:
```diff
@@ -1,2 +1,33 @@
+			> div {
+				padding: 0 var(--ee-padding-small);
+
+				> * {
+					padding-bottom: var(--ee-padding-tiny);
+				}
```
nice 👍🏻

tn3rb commented on packages/ee-components/src/Upsell/templates/BaseTemplate.tsx:
```diff
@@ -0,0 +1,45 @@
+		<div className={className}>
+			{imagePosition === 'top' && <Image src={image} />}
+			<div>
+				<Heading as='h1' className={`${prefixClassName}__main-title`}>
+					{mainTitle}
+				</Heading>
+				<Heading as='h2' className={`${prefixClassName}__subtitle`}>
+					{subtitle}
+				</Heading>
```
may want to reduce these since (AFAIK) there should ideally be only one H1 tag on a page,
so maybe use H2 and H3 (at absolute most) or H3 and H4 (or even lower) since they can be styled however we want, but would be semantically more accurate (IMHO).

nicolad replied:
thx, done

tn3rb commented on packages/ee-components/src/Upsell/templates/BaseTemplate.tsx:
```diff
@@ -0,0 +1,45 @@
+
+	return (
+		<div className={className}>
+			{imagePosition === 'top' && <Image src={image} />}
```
what about left aligned images?

```suggestion
			{(imagePosition === 'top' || imagePosition === 'left') && <Image src={image} />}
```

nicolad replied:
thanks, change applied

manzoorwanijk replied:
What about `start` and `end` instead of `left` and `right` for better RTL?

tn3rb commented on packages/ee-components/src/Upsell/templates/BaseTemplate.tsx:
```diff
@@ -0,0 +1,45 @@
+					</Button>
+				</div>
+			</div>
+			<div>{imagePosition === 'right' && <Image src={image} />}</div>
```
```suggestion
			<div>{(imagePosition === 'bottom' || imagePosition === 'right') && <Image src={image} />}</div>
```

might be better to just place the image **once** and then use flexbox to reverse the order if you want it on the other side of the text content. This of course also works regardless of orientation.

nicolad replied:
updated

tn3rb reviewed (dismissed):
Looks Great 💯 

had a couple of suggestions for possible improvements but this is absolutely great as is and any changes can be done in another PR if you prefer

nicolad commented on packages/ee-components/src/Upsell/1999047-event-management/eps/001-air walls.eps:
```diff
@@ -0,0 +1,139 @@
+%!PS-Adobe-3.0 EPSF-3.0
```
icons source:
- https://www.flaticon.com/packs/event-management
- https://www.flaticon.com/packs/event-management-1

nicolad commented on packages/ee-components/src/Upsell/stories/index.stories.tsx:
```diff
@@ -45,3 +49,12 @@ WithBaseTemplateAndTopImagePosition.args = { imagePosition: 'top', template: 'ba
 export const WithBaseTemplateAndBottomImagePosition: UpsellStory = Template.bind({});
 WithBaseTemplateAndBottomImagePosition.args = { imagePosition: 'bottom', template: 'base' };
+
+export const WithBg: UpsellStory = Template.bind({});
+WithBg.args = {
+	bgColor: 'green',
+	icon: <AttendantList />,
+	mainText: 'Sign up today',
+	mainTitle: 'Get your event attendance tracking to the next level',
+	template: 'with-bg-image',
+};
```
![image](https://user-images.githubusercontent.com/1477453/109010295-49d9c280-76b8-11eb-9de6-f863465c7889.png)

tn3rb replied:
this looks great!

nicolad replied:
the whole card is clickable, a hover/ focus state needs to be updated

tn3rb replied:
hmmmm... my initial reaction was really positive and I thought that's super cool...
but on further consideration I'm wondering if a dedicated button would be better for a11y as well as simply easier for more users to understand. I've found over the years that lots of things that seem obvious to me are not clear at all to some users and it's really best to just do things in the most predictable and painfully obvious way.

In terms of a11y it also means we can utilize all of the built-in functionality of our buttons instead of having to duplicate those efforts in other places.

nicolad replied:
created related PR:

https://github.com/eventespresso/barista/pull/738

nicolad replied:
I am not a designer, but the rationale for choosing the previous approach was that in the past project I've implemented smth similar:

https://user-images.githubusercontent.com/1477453/109162273-b02a1800-7780-11eb-88fc-9d5cc31a0252.mp4

nicolad commented:
if @tn3rb will like the SVGs I've added I will leave, if not I will remove these in the upcoming PRs;
I've added these for the demo purposes and because the theme of the icons are related to event management

nicolad commented:
if these icons will be used we can add or create credits section, as suggeste here

![image](https://user-images.githubusercontent.com/1477453/109012926-3b40da80-76bb-11eb-9cb8-b2ba8a9f6a8b.png)

manzoorwanijk reviewed (approved):
LGTM

tn3rb commented:
those SVGs look great imho and will make for some really nice marketing notices
since these will only appear in the admin, I think a footer credit is more than acceptable and I can't see @sethshoultes or @garthkoyle having an issue with that. If we needed to display credits on the frontend of the site it would be another story.
